### PR TITLE
Converting drops to custom contract

### PR DESCRIPTION
**Pull request by ACC01ADE**
## Describe Changes

Converting drops to custom contract

- Moving from drops dir to token dir
- Consolidating all drops data into single custom token contract

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] Code styles have been enforced
- [x] All Hardhat tests are passing
- [x] All Foundry tests are passing

### CODE REVIEW

File: contracts/abstract/ERC721H.sol [hunk 1]
```diff
@@ -217,7 +217,7 @@ abstract contract ERC721H is Initializable {
    * @dev This function is unreachable unless custom contract address is called directly.
    *      Please use custom payable functions for accepting native value.
    */
-  receive() external payable {
+  receive() external payable virtual {
```
Comment (alexanderattar) [inline contracts/abstract/ERC721H.sol:222]: is this so the custom contract inheriting can have a custom receive?

File: src/token/HolographDropsEditionsV1.sol [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+  function tokenURI(uint256 tokenId) public view returns (string memory) {
+    HolographERC721Interface H721 = HolographERC721Interface(holographer());
+    require(H721.exists(tokenId), "ERC721: token does not exist");
+
+    return config.metadataRenderer.tokenURI(tokenId);
+  }
+
+  event FundsReceived(address indexed source, uint256 amount);
+
+  receive() external payable override {
```
Comment (alexanderattar) [inline src/token/HolographDropsEditionsV1.sol]: oh yeah i see

File: contracts/abstract/ERC721H.sol (hunk 1 above)
Comment (ACC01ADE) [reply contracts/abstract/ERC721H.sol:222]: yep, this is to allow custom contract to override it if needed

File: contracts/drops/token/HolographDropsEditionsV1.sol [hunk 3]
```diff
@@ -245,13 +245,17 @@ contract HolographDropsEditionsV1 is NonReentrant, ERC721H, IHolographERC721Drop
     salesConfig = initializer.salesConfiguration;
 
     // TODO: Need to make sure to initialize the metadata renderer
```
Comment (alexanderattar) [inline contracts/drops/token/HolographDropsEditionsV1.sol]: The code below is handling this TODO now right? Okay to remove this comment?

File: contracts/drops/token/HolographDropsEditionsV1.sol [hunk 4]
```diff
@@ -312,6 +316,25 @@ contract HolographDropsEditionsV1 is NonReentrant, ERC721H, IHolographERC721Drop
       });
   }
 
+  function bridgeIn(
```
Comment (alexanderattar) [inline contracts/drops/token/HolographDropsEditionsV1.sol]: Will there need to be additional logic added in order to support bridging of Drop NFTs still?

File: contracts/old_drops/HolographDropCreator.sol [hunk 5]
```diff
@@ -70,7 +70,7 @@ contract HolographDropCreator is Initializable {
     uint16 royaltyBPS,
     address payable fundsRecipient,
     bytes[] memory setupCalls,
-    IMetadataRenderer metadataRenderer,
+    IMetadataRenderer sMetadataRenderer,
```
Comment (alexanderattar) [inline contracts/old_drops/HolographDropCreator.sol]: Why `sMetadataRenderer`? Maybe typo?

File: contracts/old_drops/HolographDropCreator.sol [hunk 6]
```diff
@@ -87,7 +87,7 @@ contract HolographDropCreator is Initializable {
       editionSize,
       royaltyBPS,
       setupCalls,
-      address(metadataRenderer),
+      address(sMetadataRenderer),
```
Comment (alexanderattar) [inline contracts/old_drops/HolographDropCreator.sol]: Related to above

File: deploy/01_sources.ts [hunk 7]
```diff
@@ -784,7 +784,9 @@ const func: DeployFunction = async function (hre1: HardhatRuntimeEnvironment) {
                 '0x' + web3.utils.asciiToHex('HolographGeneric').substring(2).padStart(64, '0'),
                 '0x' + web3.utils.asciiToHex('HolographERC20').substring(2).padStart(64, '0'),
                 '0x' + web3.utils.asciiToHex('HolographERC721').substring(2).padStart(64, '0'),
+                // TODO: remove this line once we depreciate old drops
                 '0x' + web3.utils.asciiToHex('HolographERC721Drop').substring(2).padStart(64, '0'),
```
Comment (alexanderattar) [inline deploy/01_sources.ts]: Will we need to remove this if we get rid of `old_drops`?

File: deploy/07_erc721drop.ts [hunk 8]
```diff
@@ -30,55 +30,155 @@ const func: DeployFunction = async function (hre1: HardhatRuntimeEnvironment) {
...
-  const futureEditionMetadataRendererAddress = await genesisDeriveFutureAddress(
+  // Deploy the HolographDropsEditionsV1 custom contract source
+  const holographDropsEditionsV1InitCode = generateInitCode(
+    [
+      'tuple(address,address,address,address,uint64,uint16,tuple(uint104,uint32,uint64,uint64,uint64,uint64,bytes32),address,bytes)',
+    ],
+    [
+      [
+        '0x0000000000000000000000000000000000000000', // holographERC721TransferHelper
+        '0x0000000000000000000000000000000000000000', // marketFilterAddress (opensea)
```
Comment (alexanderattar) [inline deploy/07_erc721drop.ts]: will we need to update this to be the actual OpenSea market filter address? I believe it's `0x000000000000AAeB6D7670E522A718067333cd4E`

File: deploy/07_erc721drop.ts [hunk 9]
```diff
@@ -30,55 +30,155 @@ const func: DeployFunction = async function (hre1: HardhatRuntimeEnvironment) {
...
-    await genesisDeployHelper(
+  if (holographDropsEditionsV1DeployedCode == '0x' || holographDropsEditionsV1DeployedCode == '') {
+    hre.deployments.log('"HolographDropsEditionsV1" bytecode not found, need to deploy"');
+    let holographDropsEditionsV1 = await genesisDeployHelper(
       hre,
       salt,
-      'EditionMetadataRenderer',
-      generateInitCode([], []), // initCode
-      futureEditionMetadataRendererAddress
+      'HolographDropsEditionsV1',
```
Comment (alexanderattar) [inline deploy/07_erc721drop.ts]: This isn't related to this particular line, but this contract name is a bit verbose. I think we can update it to just be `HolographDropsV1`. Editions are essentially just a type of Drop so it's already implied

File: deploy/07_erc721drop.ts [hunk 10]
```diff
@@ -88,7 +188,7 @@ const func: DeployFunction = async function (hre1: HardhatRuntimeEnvironment) {
       ['tuple(address,address,address,string,string,address,address,uint64,uint16,bytes[],address,bytes)', 'bool'],
       [
         [
-          futureFeeManagerAddress, // holographFeeManager
+          '0x0000000000000000000000000000000000000000', // holographFeeManager
```
Comment (alexanderattar) [inline deploy/07_erc721drop.ts]: I thought fee manager was removed from the new DropsInitializer? Or is this for the old drops contract

File: deploy/20_register_templates.ts [hunk 11]
```diff
@@ -124,57 +124,72 @@ const func: DeployFunction = async function (hre1: HardhatRuntimeEnvironment) {
...
-        true, // skipInit
-      ]
-    ) // initCode
+        '0x0000000000000000000000000000000000000000', // holographERC721TransferHelper
+        '0x0000000000000000000000000000000000000000', // marketFilterAddress (opensea)
+        deployer.address, // initialOwner
+        deployer.address, // fundsRecipient
+        0, // 1000 editions
+        1000, // 10% royalty
+        [0, 0, 0, 0, 0, 0, '0x' + '00'.repeat(32)], // setupCalls
```
Comment (alexanderattar) [inline deploy/20_register_templates.ts]: What's happening here? Is it all the params for sales config?

File: deploy/20_register_templates.ts [hunk 12]
```diff
@@ -276,6 +291,56 @@ const func: DeployFunction = async function (hre1: HardhatRuntimeEnvironment) {
   } else {
     hre.deployments.log('"HolographRoyalties" is already registered');
   }
+
+  // TODO: remove everything below this line once we depreciate old drops
```
Comment (alexanderattar) [inline deploy/20_register_templates.ts]: ahh okay

File: contracts/drops/token/HolographDropsEditionsV1.sol (hunk 4 above)
Comment (ACC01ADE) [reply contracts/drops/token/HolographDropsEditionsV1.sol]: Enforcer handles all the minting/burning so unless we have some super custom tokenURI logic that is generative and changes on bridge transfers, we do not need anything else.

File: contracts/old_drops/HolographDropCreator.sol (hunk 5 above)
Comment (ACC01ADE) [reply contracts/old_drops/HolographDropCreator.sol]: Sorry yep it’s a typo. Was doing find replace and guess this one got caught in the crossfire. Will need to fix that on next commit.

File: contracts/old_drops/HolographDropCreator.sol (hunk 6 above)
Comment (ACC01ADE) [reply contracts/old_drops/HolographDropCreator.sol]: Same as above, need to fix.

File: deploy/01_sources.ts (hunk 7 above)
Comment (ACC01ADE) [reply deploy/01_sources.ts]: Yep, there’s a few TODO comments that require removal of dead code once old_drops is removed.

File: deploy/07_erc721drop.ts (hunk 8 above)
Comment (ACC01ADE) [reply deploy/07_erc721drop.ts]: This is a deployment of source. No actual data needs to be passed on. Real addresses are supposed to be used when deploying a holographable contract though.

File: deploy/07_erc721drop.ts (hunk 9 above)
Comment (ACC01ADE) [reply deploy/07_erc721drop.ts]: Sure, we can go through the code and rename this contract.

File: deploy/07_erc721drop.ts (hunk 10 above)
Comment (ACC01ADE) [reply deploy/07_erc721drop.ts]: This is for old_drops deployment code. There’s a TODO comment above that says this should all be deleted once we remove old_drops.

File: deploy/20_register_templates.ts (hunk 11 above)
Comment (ACC01ADE) [reply deploy/20_register_templates.ts]: Yep this is sales config data. Again it’s empty cuz this is deployment of source where no actual data is being used.

File: deploy/20_register_templates.ts (hunk 12 above)
Comment (ACC01ADE) [reply deploy/20_register_templates.ts]: 👍

File: contracts/drops/token/HolographDropsEditionsV1.sol (hunk 4 above)
Comment (alexanderattar) [reply contracts/drops/token/HolographDropsEditionsV1.sol]: Okay cool so these are just hooks for extensions 👍

File: contracts/old_drops/HolographDropCreator.sol (hunk 5 above)
Comment (alexanderattar) [reply contracts/old_drops/HolographDropCreator.sol]: cool

File: deploy/07_erc721drop.ts (hunk 8 above)
Comment (alexanderattar) [reply deploy/07_erc721drop.ts]: Makes sense

File: deploy/07_erc721drop.ts (hunk 10 above)
Comment (alexanderattar) [reply deploy/07_erc721drop.ts]: okay yeah I figured it out eventually 😅

File: deploy/20_register_templates.ts (hunk 11 above)
Comment (alexanderattar) [reply deploy/20_register_templates.ts]: got it 👍

File: contracts/drops/token/HolographDropsEditionsV1.sol [hunk 13]
```diff
@@ -169,14 +192,6 @@ contract HolographDropsEditionsV1 is NonReentrant, ERC721H, IHolographERC721Drop
     _;
   }
 
-  function _presaleActive() internal view returns (bool) {
```
Comment (alexanderattar) [inline contracts/drops/token/HolographDropsEditionsV1.sol]: Why were these removed?

Comment (alexanderattar) [reply contracts/drops/token/HolographDropsEditionsV1.sol]: nevermind i see they were just moved

File: contracts/drops/oracle/DropsPriceOracleEthereum.sol [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    require(!_isInitialized(), "HOLOGRAPH: already initialized");
+    assembly {
+      sstore(_adminSlot, origin())
+    }
+    _setInitialized();
+    return Initializable.init.selector;
+  }
+
+  function convertUsdToWei(uint256 usdAmount) external view returns (uint256 weiAmount) {
+    weiAmount = usdAmount;
```
Comment (alexanderattar) [inline contracts/drops/oracle/DropsPriceOracleEthereum.sol]: isn't this where an external call needs to be made to get the conversion?

File: contracts/drops/token/HolographDropsEditionsV1.sol [hunk 15]
```diff
@@ -372,9 +379,9 @@ contract HolographDropsEditionsV1 is NonReentrant, ERC721H, IHolographERC721Drop
     onlyPublicSaleActive
     returns (uint256)
   {
-    uint256 salePrice = salesConfig.publicSalePrice;
+    uint256 salePrice = _usdToWei(salesConfig.publicSalePrice);
 
-    if (msg.value != salePrice * quantity) {
+    if (msg.value < salePrice * quantity) {
```
Comment (alexanderattar) [inline contracts/drops/token/HolographDropsEditionsV1.sol]: to allow paying over?

File: contracts/drops/token/HolographDropsEditionsV1.sol [hunk 16]
```diff
@@ -429,8 +436,9 @@ contract HolographDropsEditionsV1 is NonReentrant, ERC721H, IHolographERC721Drop
       revert Presale_MerkleNotApproved();
     }
 
-    if (msg.value != pricePerToken * quantity) {
-      revert Purchase_WrongPrice(pricePerToken * quantity);
+    uint256 weiPrice = _usdToWei(pricePerToken);
+    if (msg.value < weiPrice * quantity) {
```
Comment (alexanderattar) [inline contracts/drops/token/HolographDropsEditionsV1.sol]: maybe just to be specific use `weiPricePerToken`?

File: contracts/drops/token/HolographDropsEditionsV1.sol [hunk 17]
```diff
@@ -640,6 +648,21 @@ contract HolographDropsEditionsV1 is NonReentrant, ERC721H, IHolographERC721Drop
...
   }
 
+  function _usdToWei(uint256 amount) internal view returns (uint256 weiAmount) {
+    IDropsPriceOracle dropsPriceOracle;
+    assembly {
+      dropsPriceOracle := sload(_dropsPriceOracleSlot)
+    }
+    //require(address(dropsPriceOracle) != address(0), "price oracle not set");
+    //weiAmount = dropsPriceOracle.convertUsdToWei(amount);
+    //// TODO: remove this later when oracle is actually implemented
```
Comment (alexanderattar) [inline contracts/drops/token/HolographDropsEditionsV1.sol]: oh okay i see the conversion happens here

File: contracts/drops/oracle/DropsPriceOracleEthereum.sol (hunk 14 above)
Comment (ACC01ADE) [reply contracts/drops/oracle/DropsPriceOracleEthereum.sol]: Yep, for now just returning same value since the actual external logic is not implemented.

File: contracts/drops/token/HolographDropsEditionsV1.sol (hunk 15 above)
Comment (ACC01ADE) [reply contracts/drops/token/HolographDropsEditionsV1.sol]: We want to have an extra buffer for user to overpay since conversion might shift slightly from when the transaction was submitted. We also need to add logic to return any leftover tokens back.

File: contracts/drops/token/HolographDropsEditionsV1.sol (hunk 16 above)
Comment (ACC01ADE) [reply contracts/drops/token/HolographDropsEditionsV1.sol]: Sure

File: contracts/drops/oracle/DropsPriceOracleEthereum.sol (hunk 14 above)
Comment (alexanderattar) [reply contracts/drops/oracle/DropsPriceOracleEthereum.sol]: okay cool

File: contracts/drops/token/HolographDropsEditionsV1.sol (hunk 15 above)
Comment (alexanderattar) [reply contracts/drops/token/HolographDropsEditionsV1.sol]: okay yeah cool

File: contracts/interface/HolographTreasuryInterface.sol [hunk 18]
```diff
@@ -101,6 +101,4 @@
 
 pragma solidity 0.8.13;
 
-interface HolographTreasuryInterface {
-
-}
+interface HolographTreasuryInterface {}
```
Comment (sogoiii) [inline contracts/interface/HolographTreasuryInterface.sol]: @alexanderattar @BravoNatalie @ACC01ADE Is there anyway we can make this not happen anymore?

Comment (alexanderattar) [reply contracts/interface/HolographTreasuryInterface.sol]: I believe it's getting modified via the transpilation step and then prettier converts it back. I stopped running prettier:fix because it was messing with foundry's git submodule dependencies so i'm not sure how it happened here. probably need to format it properly via prettier during the transpilation step, but would need to look into it more

File: src/drops/filter/OwnedSubscriptionManager.sol
Comment (alexanderattar) [inline src/drops/filter/OwnedSubscriptionManager.sol:1]: is this not needed anymore?

File: test/forge/filter/OperatorFilterRegistry.sol [hunk 19]
```diff
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: MIT
 pragma solidity 0.8.13;
 
-import {IOperatorFilterRegistry} from "../../../contracts/old_drops/interfaces/IOperatorFilterRegistry.sol";
+import {IOperatorFilterRegistry} from "../interface/IOperatorFilterRegistry.sol";
```
Comment (alexanderattar) [inline test/forge/filter/OperatorFilterRegistry.sol:4]: oh i see it was moved to the test folder

File: src/drops/filter/OwnedSubscriptionManager.sol
Comment (alexanderattar) [reply src/drops/filter/OwnedSubscriptionManager.sol:1]: nevermind i see it was moved to the test dir

**Review by sogoiii — APPROVED** [review]
👍
